File: orion-platform-service/src/services/knowledge/KnowledgeService.ts
/**
 * KnowledgeService - Business logic layer for Knowledge operations (M28)
 *
 * Orchestrates space, document, and search operations through the repository layer.
 */

import { createLogger } from '../utils/logger';

const logger = pino({ name: 'LKnowledge-LService' });
import {
  KnowledgeRepository,
  KnowledgeSpace,
  CreateSpaceInput,
  UpdateSpaceInput,
  KnowledgeDoc,
  CreateDocInput,
  UpdateDocInput,
  DocVersion,
  KnowledgeSearchResult,
} from './KnowledgeRepository';

export interface DocTag {
  name: string;
  count: number;
}

export interface DocTocItem {
  id: string;
  title: string;
  parentId: string | null;
  order: number;
}

export interface SyncLog {
  id: string;
  status: 'pending' | 'running' | 'success' | 'failed';
  startedAt: Date;
  completedAt: Date | null;
  totalDocs: number;
  successDocs: number;
  failedDocs: number;
  errorMessage: string | null;
}

export class KnowledgeServiceError extends Error {
  constructor(message: string, public code: string) { super(message); this.name = 'KnowledgeServiceError'; }
}

export class KnowledgeService {
  private repository: KnowledgeRepository;
  constructor(repository: KnowledgeRepository) { this.repository = repository; }

  // ============================================================================
  // Space operations
  // ============================================================================

  async createSpace(tenantId: string, input: CreateSpaceInput): Promise<KnowledgeSpace> {
    if (!tenantId || !input.name) throw new KnowledgeServiceError('Tenant ID and space name are required', 'INVALID_INPUT');
    return this.repository.createSpace(tenantId, input);
  }

  async getSpace(id: string): Promise<KnowledgeSpace> {
    const space = await this.repository.findSpaceById(id);
    if (!space) throw new KnowledgeServiceError(`Space not found: ${id}`, 'NOT_FOUND');
    return space;
  }

  async listSpaces(tenantId: string, params?: { type?: string; search?: string; limit?: number; offset?: number }): Promise<KnowledgeSpace[]> {
    return this.repository.findAllSpaces(tenantId, params);
  }

  async updateSpace(id: string, input: UpdateSpaceInput): Promise<KnowledgeSpace> {
    const existing = await this.repository.findSpaceById(id);
    if (!existing) throw new KnowledgeServiceError(`Space not found: ${id}`, 'NOT_FOUND');
    const updated = await this.repository.updateSpace(id, input);
    if (!updated) throw new KnowledgeServiceError(`Failed to update space: ${id}`, 'UPDATE_FAILED');
    return updated;
  }

  async deleteSpace(id: string): Promise<boolean> {
    const existing = await this.repository.findSpaceById(id);
    if (!existing) throw new KnowledgeServiceError(`Space not found: ${id}`, 'NOT_FOUND');
    return this.repository.deleteSpace(id);
  }

  // ============================================================================
  // Document operations
  // ============================================================================

  async createDoc(tenantId: string, input: CreateDocInput): Promise<KnowledgeDoc> {
    if (!tenantId || !input.title || !input.content || !input.space_id) {
      throw new KnowledgeServiceError('Tenant ID, title, content, and space_id are required', 'INVALID_INPUT');
    }

    // Verify space exists and belongs to tenant
    const space = await this.repository.findSpaceById(input.space_id);
    if (!space || space.tenant_id !== tenantId) {
      throw new KnowledgeServiceError(`Space not found: ${input.space_id}`, 'NOT_FOUND');
    }

    return this.repository.createDoc(tenantId, input);
  }

  async getDoc(id: string): Promise<KnowledgeDoc> {
    const doc = await this.repository.findDocById(id);
    if (!doc) throw new KnowledgeServiceError(`Document not found: ${id}`, 'NOT_FOUND');
    return doc;
  }

  async listDocs(tenantId: string, params?: { spaceId?: string; status?: string; tag?: string; search?: string; type?: string; limit?: number; offset?: number }): Promise<KnowledgeDoc[]> {
    return this.repository.findAllDocs(tenantId, params);
  }

  async updateDoc(id: string, input: UpdateDocInput): Promise<KnowledgeDoc> {
    const existing = await this.repository.findDocById(id);
    if (!existing) throw new KnowledgeServiceError(`Document not found: ${id}`, 'NOT_FOUND');
    const updated = await this.repository.updateDoc(id, input);
    if (!updated) throw new KnowledgeServiceError(`Failed to update document: ${id}`, 'UPDATE_FAILED');
    return updated;
  }

  async deleteDoc(id: string): Promise<boolean> {
    const existing = await this.repository.findDocById(id);
    if (!existing) throw new KnowledgeServiceError(`Document not found: ${id}`, 'NOT_FOUND');
    return this.repository.deleteDoc(id);
  }

  // ============================================================================
  // Version operations
  // ============================================================================

  async getDocVersions(docId: string): Promise<DocVersion[]> {
    // Verify doc exists
    const doc = await this.repository.findDocById(docId);
    if (!doc) throw new KnowledgeServiceError(`Document not found: ${docId}`, 'NOT_FOUND');
    return this.repository.getDocVersions(docId);
  }

  // ============================================================================
  // Search / RAG
  // ============================================================================

  async search(tenantId: string, query: string, params?: { spaceId?: string; limit?: number }): Promise<KnowledgeSearchResult[]> {
    if (!query || query.trim().length === 0) {
      throw new KnowledgeServiceError('Search query cannot be empty', 'INVALID_INPUT');
    }
    return this.repository.search(tenantId, query.trim(), params);
  }

  async retrieve(tenantId: string, query: string, params?: { spaceId?: string; topK?: number }): Promise<KnowledgeSearchResult[]> {
    // RAG retrieve: semantic or text search for retrieval purposes
    return this.repository.search(tenantId, query, { spaceId: params?.spaceId, limit: params?.topK || 5 });
  }

  // ============================================================================
  // Document Center (type=docs)
  // ============================================================================

  async listDocsByType(tenantId: string, params?: { tag?: string; search?: string; limit?: number; offset?: number }): Promise<KnowledgeDoc[]> {
    return this.repository.findAllDocs(tenantId, { ...params, type: 'docs' });
  }

  async getDocTags(tenantId: string): Promise<DocTag[]> {
    const docs = await this.repository.findAllDocs(tenantId, { type: 'docs', limit: 1000 });
    const tagMap = new Map<string, number>();

    for (const doc of docs) {
      for (const tag of doc.tags || []) {
        tagMap.set(tag, (tagMap.get(tag) || 0) + 1);
      }
    }

    return Array.from(tagMap.entries())
      .map(([name, count]) => ({ name, count }))
      .sort((a, b) => b.count - a.count);
  }

  async getDocToc(tenantId: string): Promise<DocTocItem[]> {
    const docs = await this.repository.findAllDocs(tenantId, { type: 'docs', status: 'published', limit: 200 });
    return docs.map((doc, index) => ({
      id: doc.id,
      title: doc.title,
      parentId: null,
      order: index,
    }));
  }

  async triggerSync(tenantId: string, source?: string): Promise<SyncLog> {
    // Generate a sync log entry (in production this would trigger an async job)
    const syncLog: SyncLog = {
      id: `sync-${Date.now()}`,
      status: 'success',
      startedAt: new Date(),
      completedAt: new Date(),
      totalDocs: 0,
      successDocs: 0,
      failedDocs: 0,
      errorMessage: null,
    };

    // In production: trigger async document sync from external source
    // For now, return mock success
    logger.info(`[DocumentCenter] Sync triggered for tenant ${tenantId}, source: ${source || 'manual'}`);

    return syncLog;
  }

  async getSyncLogs(tenantId: string, limit: number = 10): Promise<SyncLog[]> {
    // Return mock sync logs (in production, store in database)
    const logs: SyncLog[] = [];
    const now = new Date();

    for (let i = 0; i < Math.min(limit, 5); i++) {
      logs.push({
        id: `sync-${now.getTime() - i * 3600000}`,
        status: i === 0 ? 'success' : 'success',
        startedAt: new Date(now.getTime() - i * 3600000),
        completedAt: new Date(now.getTime() - i * 3600000 + 5000),
        totalDocs: Math.floor(Math.random() * 50) + 10,
        successDocs: Math.floor(Math.random() * 45) + 10,
        failedDocs: 0,
        errorMessage: null,
      });
    }

    return logs;
  }
}

File: orion-knowledge-svc-py/app/services/knowledge_service.py
"""Knowledge service - business logic for knowledge operations."""

from typing import Optional
from uuid import UUID, uuid4

from ..models.schemas import (
    CreateDocRequest,
    CreateSpaceRequest,
    DocVersion,
    KnowledgeDoc,
    KnowledgeSpace,
    PaginationParams,
    SearchRequest,
    SearchResult,
    UpdateDocRequest,
)


class KnowledgeService:
    """Service for managing knowledge spaces and documents."""

    def __init__(self):
        self._spaces: dict[UUID, KnowledgeSpace] = {}
        self._docs: dict[UUID, KnowledgeDoc] = {}
        self._versions: dict[UUID, list[DocVersion]] = {}

    def create_space(self, tenant_id: str, req: CreateSpaceRequest) -> KnowledgeSpace:
        space = KnowledgeSpace(
            id=uuid4(),
            tenant_id=tenant_id,
            name=req.name,
            description=req.description,
            type=req.type,
        )
        self._spaces[space.id] = space
        return space

    def get_space(self, space_id: UUID) -> Optional[KnowledgeSpace]:
        return self._spaces.get(space_id)

    def list_spaces(self, tenant_id: str, params: PaginationParams) -> list[KnowledgeSpace]:
        spaces = [s for s in self._spaces.values() if s.tenant_id == tenant_id]
        spaces.sort(key=lambda s: s.created_at, reverse=True)
        offset = (params.page - 1) * params.page_size
        return spaces[offset:offset + params.page_size]

    def create_doc(self, tenant_id: str, space_id: UUID, req: CreateDocRequest) -> KnowledgeDoc:
        space = self._spaces.get(space_id)
        if not space:
            raise ValueError(f"Space not found: {space_id}")

        doc = KnowledgeDoc(
            id=uuid4(),
            space_id=space_id,
            tenant_id=tenant_id,
            title=req.title,
            content=req.content,
            tags=req.tags,
        )
        self._docs[doc.id] = doc
        self._versions[doc.id] = [
            DocVersion(id=uuid4(), doc_id=doc.id, version=1, content=req.content)
        ]
        space.doc_count += 1
        return doc

    def get_doc(self, doc_id: UUID) -> Optional[KnowledgeDoc]:
        return self._docs.get(doc_id)

    def list_docs(self, space_id: UUID, params: PaginationParams) -> list[KnowledgeDoc]:
        docs = [d for d in self._docs.values() if d.space_id == space_id]
        docs.sort(key=lambda d: d.updated_at, reverse=True)
        offset = (params.page - 1) * params.page_size
        return docs[offset:offset + params.page_size]

    def update_doc(self, doc_id: UUID, req: UpdateDocRequest) -> KnowledgeDoc:
        doc = self._docs.get(doc_id)
        if not doc:
            raise ValueError(f"Document not found: {doc_id}")

        if req.title is not None:
            doc.title = req.title
        if req.content is not None:
            doc.content = req.content
            doc.version += 1
            self._versions[doc_id].append(
                DocVersion(id=uuid4(), doc_id=doc_id, version=doc.version, content=req.content)
            )
        if req.tags is not None:
            doc.tags = req.tags
        doc.updated_at = __import__("datetime").datetime.utcnow()
        return doc

    def delete_doc(self, doc_id: UUID) -> bool:
        doc = self._docs.pop(doc_id, None)
        if not doc:
            return False
        self._versions.pop(doc_id, None)
        space = self._spaces.get(doc.space_id)
        if space:
            space.doc_count = max(0, space.doc_count - 1)
        return True

    def search(self, tenant_id: str, req: SearchRequest) -> list[SearchResult]:
        results = []
        query_lower = req.query.lower()
        for doc in self._docs.values():
            if doc.tenant_id != tenant_id:
                continue
            if req.space_id and doc.space_id != req.space_id:
                continue
            if query_lower in doc.title.lower() or query_lower in doc.content.lower():
                snippet_start = doc.content.lower().find(query_lower)
                start = max(0, snippet_start - 50)
                end = min(len(doc.content), snippet_start + len(req.query) + 50)
                results.append(
                    SearchResult(
                        doc_id=doc.id,
                        title=doc.title,
                        snippet=doc.content[start:end],
                        score=1.0,
                        space_id=doc.space_id,
                    )
                )
        results.sort(key=lambda r: r.score, reverse=True)
        return results[:req.limit]

    def get_doc_versions(self, doc_id: UUID) -> list[DocVersion]:
        return self._versions.get(doc_id, [])
